File: day16/day16.py
#!/usr/bin/env python3

HEXDIGITS = "0123456789ABCDEF"

class Bitstream(object):
    def __init__(self, hexstr):
        self.hexstr = hexstr
        self.hexpos = 0
        self.rawdata = 0
        self.avail = 0
        self.pos = 0

    def get(self, bits):
        while self.avail < bits:
            if self.hexpos >= len(self.hexstr):
                raise RuntimeError("EOF")

            hexval = HEXDIGITS.find(self.hexstr[self.hexpos])
            if hexval < 0:
                continue

            self.rawdata = self.rawdata << 4 | hexval
            self.hexpos += 1
            self.avail += 4

        self.avail -= bits
        self.pos += bits

        value = self.rawdata >> self.avail
        self.rawdata &= (1<<self.avail) - 1
        return value

def parse(stream):
    ver = stream.get(3)
    typ = stream.get(3)
    if typ == 4:
        value = 0
        while True:
            group = stream.get(5)
            value = value << 4 | (group & 15)
            if group & 16 == 0:
                break

        return (ver, typ, value)
    elif stream.get(1) == 0:
        end = stream.get(15)
        end += stream.pos
        lst = []
        while stream.pos < end:
            lst.append(parse(stream))
        return (ver, typ, tuple(lst))
    else:
        count = stream.get(11)
        lst = tuple(parse(stream) for x in range(count))
        return (ver, typ, lst)

def interp1(tree):
    ver, typ, children = tree
    if typ == 4:
        return ver
    else:
        return ver + sum(interp1(x) for x in children)

def interp2(tree):
    ver, typ, children = tree
    if typ == 0:                # sum
        return sum(interp2(x) for x in children)
    elif typ == 1:              # product
        m = 1
        for x in children:
            m *= interp2(x)
        return m
    elif typ == 2:              # minimum
        return min(interp2(x) for x in children)
    elif typ == 3:              # maximum
        return max(interp2(x) for x in children)
    elif typ == 4:              # value
        return children
    elif typ == 5:              # greater than
        a = interp2(children[0])
        b = interp2(children[1])
        if a > b:
            return 1
        else:
            return 0
    elif typ == 6:              # less than
        a = interp2(children[0])
        b = interp2(children[1])
        if a < b:
            return 1
        else:
            return 0
    elif typ == 7:              # equal to
        a = interp2(children[0])
        b = interp2(children[1])
        if a == b:
            return 1
        else:
            return 0
    else:
        raise RuntimeError("unknown type")

def tolisp(tree):
    var, typ, children = tree
    if typ == 0:                # sum
        return "(+ " + " ".join(tolisp(x) for x in children) + ")"
    elif typ == 1:              # product
        return "(* " + " ".join(tolisp(x) for x in children) + ")"
    elif typ == 2:              # minimum
        return "(min " + " ".join(tolisp(x) for x in children) + ")"
    elif typ == 3:              # maximum
        return "(max " + " ".join(tolisp(x) for x in children) + ")"
    elif typ == 4:              # value
        return str(children)
    elif typ == 5:              # greater than
        return "(if (> " + " ".join(tolisp(x) for x in children) + ") 1 0)"
    elif typ == 6:              # less than
        return "(if (< " + " ".join(tolisp(x) for x in children) + ") 1 0)"
    elif typ == 7:              # equal to
        return "(if (= " + " ".join(tolisp(x) for x in children) + ") 1 0)"
    else:
        raise RuntimeError("unknown type")

import sys

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read()

        for line in txt.splitlines():
            program = parse(Bitstream(line))
            print("Part1:", interp1(program))
            print("Part2:", interp2(program))
            #print("Lisp:", tolisp(program))
    except FileNotFoundError:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)

File: day16/day16.cpp
#include <cctype>
#include <fstream>
#include <memory>
#include <numeric>
#include <vector>
#include <fmt/format.h>

using namespace std;

class Bitstream
{
public:
	explicit Bitstream(const string& str)
		: cur(str.cbegin()),
		  end(str.cend()),
		  rawdata(0),
		  available(0),
		  pos(0)
	{}

	unsigned get(int needed)
	{
		while (available < needed)
		{
			if (cur == end)
			{
				fmt::print(stderr, "EOF\n");
				exit(1);
			}
			int val = *cur++;
			if (isdigit(val))
			{
				val -= '0';
			}
			else if ('A' <= val && val <= 'F')
			{
				val -= 'A' - 10;
			}
			else
			{
				continue;
			}
			rawdata = rawdata << 4 | val;
			available += 4;
		}

		pos += needed;
		available -= needed;

		unsigned value = rawdata >> available;
		rawdata &= (1U<<available) - 1;
		return value;
	}

	size_t getPos() const
	{
		return pos;
	}

private:
	string::const_iterator cur;
	string::const_iterator end;
	unsigned rawdata;
	int available;
	size_t pos;
};

struct Node
{
	int version;
	int type;
	uint64_t value;
	vector<unique_ptr<Node>> children;
};

static unique_ptr<Node> parse(Bitstream& bs)
{
	auto node = make_unique<Node>();

	node->version = bs.get(3);
	node->type = bs.get(3);
	if (node->type == 4)
	{
		uint64_t value = 0;
		unsigned group;
		do
		{
			group = bs.get(5);
			value = value << 4 | (group & 0xF);
		} while (group & 0x10);
		node->value = value;
	}
	else if (bs.get(1))
	{
		size_t count = bs.get(11);
		while (count-->0)
		{
			node->children.push_back(parse(bs));
		}
	}
	else
	{
		size_t end = bs.get(15);
		end += bs.getPos();
		while (bs.getPos() < end)
		{
			node->children.push_back(parse(bs));
		}
	}
	return node;
}

static int interp1(const Node& node)
{
	if (node.type == 4)
	{
		return node.version;
	}

	return accumulate(
		node.children.cbegin(),
		node.children.cend(),
		node.version,
		[](auto acc, const auto& node)
		{
			return acc + interp1(*node);
		});
}

static uint64_t interp2(const Node& node)
{
	uint64_t value;
	switch (node.type)
	{
	case 0:
		value = accumulate(
			node.children.cbegin(),
			node.children.cend(),
			0ULL,
			[](auto acc, const auto& node)
			{
				return acc + interp2(*node);
			});
		break;

	case 1:
		value = accumulate(
			node.children.cbegin(),
			node.children.cend(),
			1ULL,
			[](auto acc, const auto& node)
			{
				return acc * interp2(*node);
			});
		break;

	case 2:
		value = UINT64_MAX;
		for (auto& n: node.children)
		{
			uint64_t tmp = interp2(*n);
			if (value > tmp)
			{
				value = tmp;
			}
		}
		break;

	case 3:
		value = 0;
		for (auto& n: node.children)
		{
			uint64_t tmp = interp2(*n);
			if (value < tmp)
			{
				value = tmp;
			}
		}
		break;

	case 4:
		value = node.value;
		break;

	case 5:
		value = interp2(*node.children[0]) > interp2(*node.children[1]);
		break;

	case 6:
		value = interp2(*node.children[0]) < interp2(*node.children[1]);
		break;

	case 7:
		value = interp2(*node.children[0]) == interp2(*node.children[1]);
		break;

	default:
		fmt::print(stderr, "Unknown Type ID {}\n", node.type);
		exit(1);
	}
	return value;
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		fmt::print(stderr, "Usage: {} <filename>\n", argv[0]);
		return 1;
	}

	ifstream in(argv[1]);
	if (!in)
	{
		fmt::print(stderr, "Cannot open {}\n", argv[1]);
		return 1;
	}

	string line;
	while (getline(in, line))
	{
		auto bs = Bitstream(line);
		auto program = parse(bs);
		fmt::print("Part1: {}\n", interp1(*program));
		fmt::print("Part2: {}\n", interp2(*program));
	}
	in.close();
	return 0;
}
